### PR TITLE
[Merged by Bors] - sync: notify tortoise about certified block after downloading a block

### LINKED ISSUE #5519: actively ask peer for certificates if you don't have one locally after hare terminated
we have two protocol parameters in layers:
- zdist . determines how long hare may take to terminate. right now it is set to 2 layers, in each layer we can have 2 hare session. 4 session max
- hdist. determines how long honest node votes according to the hare result. set 10 layers, as it also control when tortoise switches to full mode.

once layer is past zdist and we don't have certificate locally we need to ask peer for it more actively then we are doing it now.
otherwise verifying tortoise can't make progress and node will observe errors as below:

```
2024-01-30T20:31:49.015+0100	ERROR	fd68b.blockGenerator	failed to process hare output	{"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "blockGenerator", "layer_id": 57738, "errmsg": "get prev mesh hash database: not found layer 57737", "name": "blockGenerator"}
2024-01-30T20:36:49.014+0100	INFO	fd68b.blockGenerator	node has different mesh hash from majority, will download block instead	{"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "blockGenerator", "layer_id": 57739, "errmsg": "node has different mesh hash from majority: majority 3e699c0cdf, node 0000000000", "name": "blockGenerator"}
```  

in my example node got "stuck" on layer 57736 (see hare: false field)

```
2024-01-30T20:42:38.291+0100	DEBUG	fd68b.mesh	consensus results	{"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "mesh", "sessionId": "d50a3bd9-149c-48e0-bb9b-1cc4972278d7", "layer_id": 57738, "results": [{"layer": 57736, "opinion": "66647ef527", "verified": false, "blocks": [{"id": "e6912f697e", "layer": 57736, "height": 112842, "valid": false, "invalid": false, "hare": false, "data": true, "local": false}]}, {"layer": 57737, "opinion": "3c02e2432a", "verified": false, "blocks": [{"id": "12ebcd2acf", "layer": 57737, "height": 112957, "valid": false, "invalid": false, "hare": true, "data": true, "local": false}]}, {"layer": 57738, "opinion": "3dc9a2f565", "verified": false, "blocks": [{"id": "e15443d4a4", "layer": 57738, "height": 112842, "valid": false, "invalid": false, "hare": true, "data": true, "local": false}]}], "name": "mesh"}
```

#### solution

we should have a loop that periodically asks a random good peer for a certificate in the layer within [current - hdist, current - zdist]. current is the layer according the clock.

it can be implemented as its own self-contained module, (e.g certsync) that will wake up every 10-30s or on notification, and ask random peer. it can be also more sophisticated and introduce some backoff for previously asked peers.

mentioned notification can be used similar to the notification to blocksync, when mesh notices that there is a block in the layer without certificate (hare will be false, data will be true).

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/5519

tortoise requires block to exist in state (added either as a reference on ballot or direct call to OnBlock) 
when OnHareOutput is called. this wasn't correctly handled in syncer and fixed in this change.

### CODE REVIEW

File: mesh/mesh.go [hunk 1]
```diff
@@ -140,6 +140,11 @@ func (msh *Mesh) recoverFromDB(latest types.LayerID) {
 		log.Stringer("processed", msh.ProcessedLayer()))
 }
 
+// Tortoise returns the tortoise instance.
+func (mesh *Mesh) Tortoise() system.Tortoise {
+	return mesh.trtl
+}
```
Comment (fasmat) [inline mesh/mesh.go]: This hides the dependency of the `Syncer` through the dependency on the `Mesh`.

Would it make sense to instead make the `Syncer` depend on tortoise directly?

**Review by fasmat — APPROVED** [review]

File: mesh/mesh.go (hunk 1 above)
Comment (dshulyak) [reply mesh/mesh.go]: you are right, i will change that

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**